Credit -> KHUSHEE KAPOOR

In [15]:
# importing the libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn  as sns

In [16]:
# reading the dataset
df = pd.read_csv('./data/AusDataForRainPred.csv')
vifdf = pd.read_csv('./data/AusDataForRainPred.csv')

In [17]:
# viewing the first 10 rows
df.head(10)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5  2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6  2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7  2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8  2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   
9  2008-12-10   Albury     13.1     30.1       1.4          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   
5         WNW           56.0          W  ...        55.0         23.0   
6           W           50.0         SW  ...        49.0         19.0   
7           W           35.0        SSE  ...        48.0         19.0   
8         NNW           80.0         SE  ...        42.0          9.0   
9           W           28.0          S  ...        58.0         27.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   
5       1009.2       1005.4       NaN       NaN     20.6     28.9         No   
6       1009.6       1008.2       1.0       NaN     18.1     24.6         No   
7       1013.4       1010.1       NaN       NaN     16.3     25.5         No   
8       1008.9       1003.6       NaN       NaN     18.3     30.2         No   
9       1007.0       1005.7       NaN       NaN     20.1     28.2        Yes   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  
5            No  
6            No  
7            No  
8           Yes  
9            No  

[10 rows x 23 columns]

In [18]:
# viewing the dimensions of the dataframe
df.shape

(145460, 23)

In [19]:
# checking for missing values
df.isnull().sum()

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64

In [20]:
# dropping the missing values
df = df.dropna()
vifdf = vifdf.dropna()

In [21]:
# encoding the column
df['RainToday'] = df['RainToday'].map({'No':0, 'Yes':1})
vifdf['RainToday'] = vifdf['RainToday'].map({'No':0, 'Yes':1})

In [22]:
# selecting the categorical columns
categorical_columns = ['Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm']

# encoding the values
for column in categorical_columns:
    
    # in original dataframe
    catdf = pd.get_dummies(df[column], prefix=column)    
    df = pd.merge(left=df, right=catdf, left_index=True, right_index=True)    
    df = df.drop(columns=column)
    
    # in duplicate dataframe
    catdf = pd.get_dummies(vifdf[column], prefix=column)    
    vifdf = pd.merge(left=vifdf, right=catdf, left_index=True, right_index=True)    
    vifdf = vifdf.drop(columns=column)

In [23]:
# splitting the data into independent and dependent variables
x = df.drop(columns=['Date', 'RainTomorrow'])
y = df['RainTomorrow']

# diving the dataset into training and testing sets
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=105)

In [24]:
# building the model
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()

# training the model
logreg.fit(x_train, y_train)

# printing the accuracy
print(str.format('Accuracy: {:.2f}%', logreg.score(x_test, y_test)*100))

Accuracy: 84.93%


c:\Users\USER\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [25]:
# importing the variance_inflation_factor() function
from statsmodels.stats.outliers_influence import variance_inflation_factor

# extracting the numerical columns
cols = [cname for cname in vifdf.columns if vifdf[cname].dtype in ['int64', 'float64']]
data = vifdf[cols]
  
# VIF dataframe
vif_data = pd.DataFrame()
vif_data['Feature'] = data.columns
  
# calculating VIF for each feature
vif_data['VIF']= [variance_inflation_factor(data.values, i) for i in range(len(data.columns))]

# printing the VIF of each feature
print(vif_data)

          Feature            VIF
0         MinTemp      58.284148
1         MaxTemp     609.736465
2        Rainfall       1.627144
3     Evaporation       7.220027
4        Sunshine      17.318912
5   WindGustSpeed      26.969533
6    WindSpeed9am       8.403752
7    WindSpeed3pm      13.776204
8     Humidity9am      61.146908
9     Humidity3pm      47.805246
10    Pressure9am  432633.791186
11    Pressure3pm  430736.444717
12       Cloud9am       7.362031
13       Cloud3pm       8.322534
14        Temp9am     210.815320
15        Temp3pm     674.785736
16      RainToday       2.227511


In [28]:
# engineering the new features
vifdf['Range'] = vifdf['MaxTemp'] - vifdf['MinTemp']
vifdf['WindSpeed'] = vifdf['WindSpeed3pm'] - vifdf['WindSpeed9am']
vifdf['Humidity'] = vifdf['Humidity3pm'] - vifdf['Humidity9am']
vifdf['Pressure'] = vifdf['Pressure3pm'] - vifdf['Pressure9am']
vifdf['Cloud'] = vifdf['Cloud3pm'] - vifdf['Cloud9am']
vifdf['Temp'] = vifdf['Temp3pm'] - vifdf['Temp9am']

# dropping the original columns
vifdf = vifdf.drop(columns=['MaxTemp', 'MinTemp', 'WindSpeed3pm', 'WindSpeed9am', 'Humidity3pm', 'Humidity9am', 'Pressure3pm',
                     'Pressure9am', 'Cloud3pm', 'Cloud9am', 'Temp3pm', 'Temp9am'])

KeyError: 'MaxTemp'

In [29]:
# importing the variance_inflation_factor() function
from statsmodels.stats.outliers_influence import variance_inflation_factor

# extracting the numerical columns
cols = [cname for cname in vifdf.columns if vifdf[cname].dtype in ['int64', 'float64']]
data = vifdf[cols]
  
# VIF dataframe
vif_data = pd.DataFrame()
vif_data['Feature'] = data.columns
  
# calculating VIF for each feature
vif_data['VIF']= [variance_inflation_factor(data.values, i) for i in range(len(data.columns))]

# printing the VIF of each feature
print(vif_data)

          Feature        VIF
0        Rainfall   1.591422
1     Evaporation   4.373922
2        Sunshine   8.469864
3   WindGustSpeed   6.641747
4       RainToday   2.061839
5           Range  19.539385
6       WindSpeed   1.371614
7        Humidity   5.575934
8        Pressure   3.304605
9           Cloud   1.077251
10           Temp  13.968278


As we can see, we have significantly reduced the VIF for many columns by engineering new features and removing the original ones.

In [30]:
# splitting the data into independent and dependent variables
x = vifdf.drop(columns=['Date', 'Sunshine', 'WindGustSpeed', 'Range', 'Temp', 'RainTomorrow'])
y = vifdf['RainTomorrow']

# diving the dataset into training and testing sets
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=105)

In [31]:
# building the model
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()

# training the model
logreg.fit(x_train, y_train)

# printing the accuracy
print(str.format('Accuracy: {:.2f}%', logreg.score(x_test, y_test)*100))

Accuracy: 81.25%


c:\Users\USER\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


As we can see, the accuracy of the model is 81.25%. The accuracy has slightly dropped in comparison to the model built using the original dataset. However, the model is now much more stable because we have removed the multicollinearity, so the variance of the coefficient estimate is stable and is not very sensitive to minor changes in the model.